### PR TITLE
[11.x] Introduce Schedule Grouping

### PR DESCRIPTION
**Pull request by istiak-tridip**
### Description

This PR introduces a new feature allowing related schedules to be grouped, similar to the `Route::group` method.

With this addition, developers can configure multiple schedules collectively, reducing the need for repetitive configuration of individual schedules.

Looking forward to your feedback. Thanks for all the great work that you guys do! 😍

### Current Behavior
```php
Schedule::command('command-one')->everyMinute()->runInBackground()->withoutOverlapping();
Schedule::command('command-two')->everyMinute()->runInBackground()->withoutOverlapping();
Schedule::command('command-three')->everyMinute()->runInBackground()->withoutOverlapping();
```

### With This PR
```php
Schedule::group()
    ->everyMinute()
    ->runInBackground()
    ->withoutOverlapping()
    ->schedules(function () {
        Schedule::command('command-one');
        Schedule::command('command-two');
        Schedule::command('command-three');
    });
```

Individual schedules can override group configurations as needed:
```php
Schedule::group()
    ->everyMinute()
    ->runInBackground()
    ->withoutOverlapping()
    ->schedules(function () {
        Schedule::command('command-one');
        Schedule::command('command-two');
        Schedule::command('command-three')->everyTenMinutes();  // Override the group's cron expression
    });
```

Nested grouping is also supported:
```php
Schedule::group()
    ->runInBackground()
    ->withoutOverlapping()
    ->schedules(function () {
        Schedule::group()->everyMinute()->schedules(function () {
            Schedule::command('command-one');
            Schedule::command('command-two');
        });

        Schedule::group()->everyTenMinutes()->schedules(function () {
            Schedule::command('command-three');
            Schedule::command('command-four');
        });
    });
```
----
Past similar proposals: #49832 #48046

### CODE REVIEW

**Comment by decadence** [comment]
Hope this will be merged. I think about this every time I call `runInBackground` for each event.

**Comment by rodrigopedra** [comment]
I'd prefer a syntax like this:

```php
Schedule::group(function () {
    Schedule::command('command-one');
    Schedule::command('command-two');
})
    ->runInBackground()
    ->withoutOverlapping()
    ->everyMinute();
```

**Comment by istiak-tridip** [comment]
Thanks for the suggestion, @rodrigopedra!

In my view, the suggested syntax makes the chaining feel a bit less clean and slightly harder to follow. Keeping these configurations at the start helps highlight the group’s behavior, improving readability and clarity overall.

That said, if there’s strong preference for this approach, I’m definitely open to revisiting the implementation.

**Comment by rodrigopedra** [comment]
My preference is that it would be more inline with the route group feature, mentioned as inspiration for this feature.

Also, in my opinion, it is more declarative. 

But let's wait for the maintainer's opinion.

Other than that, I think this is a great addition!

**Comment by Rizky92** [comment]
> I'd prefer a syntax like this:
> 
> ```
> Schedule::group(function () {
>     Schedule::command('command-one');
>     Schedule::command('command-two');
> })
>     ->runInBackground()
>     ->withoutOverlapping()
>     ->everyMinute();
> ```

I agree! This is more inlined with how Laravel handles fluent route method.

**Comment by istiak-tridip** [comment]
@Rizky92 @rodrigopedra I looked into the route group feature again, and it seems that using the syntax you suggested doesn’t apply the group options to the routes.

For example:
```php
Route::group([], function () {
    Route::get('/test', fn () => 'Admin Dashboard')->name('home');
})
    ->name('admin.') // doesn’t affect routes
    ->prefix('/admin'); // doesn’t affect routes
```

Instead, it needs to be structured like this:
```php
Route::name('admin.')
    ->prefix('/admin')
    ->group(function () {
        Route::get('/test', fn () => 'Admin Dashboard')->name('home');
    });
```

This setup aligns with my suggested syntax. But let’s see what the maintainers think before making any changes. 🙏

File: tests/Integration/Console/Scheduling/ScheduleGroupTest.php [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+declare(strict_types=1);
+
+namespace Illuminate\Tests\Integration\Console\Scheduling;
+
+use Illuminate\Console\Scheduling\Schedule as ScheduleClass;
+use Illuminate\Support\Facades\Schedule;
+use Orchestra\Testbench\TestCase;
+use PHPUnit\Framework\Attributes\DataProvider;
+
+final class ScheduleGroupTest extends TestCase
```
Comment (stevebauman) [inline tests/Integration/Console/Scheduling/ScheduleGroupTest.php:12]: Final classes aren't used for framework test cases.

```suggestion
class ScheduleGroupTest extends TestCase
```

File: src/Illuminate/Console/Scheduling/ScheduleGroup.php [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        }
+
+        // Merge the reject callbacks into the event.
+        foreach ($this->rejects as $reject) {
+            $event->skip($reject);
+        }
+
+        /**
+         * Loop through the attributes and only set the ones that were set on the group.
+         */
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/ScheduleGroup.php:114]: Inconsistent comment format:

```suggestion
        // Loop through the attributes and only set the ones that were set on the group.
```

File: src/Illuminate/Console/Scheduling/ScheduleGroup.php [hunk 3]
```diff
@@ -0,0 +1,221 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Illuminate\Console\Scheduling;
+
+use Closure;
+use DateTimeZone;
+
+class ScheduleGroup
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/ScheduleGroup.php:10]: Class is missing doc blocks on some methods, which must be implemented on all methods (including constructor).

**Review by stevebauman — COMMENTED** [review]
I like this idea! 👍

File: src/Illuminate/Console/Scheduling/ScheduleGroup.php (hunk 3 above)
Comment (istiak-tridip) [reply src/Illuminate/Console/Scheduling/ScheduleGroup.php:10]: I have added the missing doc blocks. Thanks 👍

**Comment by taylorotwell** [comment]
Really would love to finally get this in the framework in some form. I am in agreement that consistency with the documented Route group syntax would be awesome where `group` is at the end and accepts a Closure.

**Comment by istiak-tridip** [comment]
@taylorotwell I can think of two ways to implement the `Route::group`-style syntax:

1. Implement it similarly to the `Router` class, using `__call` to proxy relevant method calls to a "Schedule Grouper." This keeps event attributes separate from the `Schedule` class, as they are currently, but doesn’t provide native IDE auto-completion.

2. Alternatively, duplicate the relevant methods from the `Scheduling\Event` class into the `Schedule` class to ensure auto-completion support.

Let me know which direction you'd prefer!

**Comment by morloderex** [comment]
> @taylorotwell I can think of two ways to implement the `Route::group`-style syntax:
> 
> 1. Implement it similarly to the `Router` class, using `__call` to proxy relevant method calls to a "Schedule Grouper." This keeps event attributes separate from the `Schedule` class, as they are now, maintaining a clean structure but lacking native IDE auto-completion.
> 
> 2. Alternatively, duplicate the relevant methods from the `Scheduling\Event` class into the `Schedule` class to ensure auto-completion support.
> 
> Let me know which direction you'd prefer!

How about extracting the relevant method to a trait and use that trait on both classes? That way, you get auto completion without duplicating the methods?

**Comment by istiak-tridip** [comment]
@morloderex Thanks for the reminder! That’s actually what I had planned to do; I just forgot to mention it. I’ve updated my previous reply to reflect this.

**Comment by morloderex** [comment]
> @taylorotwell I've taken a closer look at the codebase to explore implementing the syntax through the second method if we go that route. I've identified some potential side effects with this approach.
> 
> For example, end users could write:
> 
> ```
> Schedule::runInBackground()->command('command-one');
> ```
> 
> While that seems fine (similar to routes), issues arise with the `callback event`. The `Schedule::call` returns a `CallbackEvent` instance, which overrides some methods of the parent `Event` class.
> 
> For instance, `runInBackground` is prohibited for `callback event`:
> 
> ```
> Schedule::call('command-one')->runInBackground(); // Throws `RuntimeException`
> ```
> 
> Allowing `runInBackground` before knowing the event type prevents us from replicating this behavior.
> 
> A solution could be to only allow `group` to be called after event attribute methods or to add a check in `Schedule::call `before creating the event. Both options add complexity.
> 
> What do you think?

Wondering if it would work by just making `runInBackground` return somekind of shadow object which is just to allow chaining in this case then? And not actually make it throw the exception anymore. And then add a note in the docs for this specific method chain?

**Comment by istiak-tridip** [comment]
@morloderex That approach is possible, but it would completely change the current behavior, which may cause existing tests to fail and go against end-users' existing expectations. It might also be considered a breaking change, though I'm not entirely sure.

For now, it may be best to wait for feedback from @taylorotwell or other maintainers.

**Comment by istiak-tridip** [comment]
@stevebauman @taylorotwell I’ve updated the implementation to follow `Route::group` syntax, please review 🙏.

Here’s an example of the new `Schedule` group usage:
```php
Schedule::everyMinute()
    ->runInBackground()
    ->withoutOverlapping()
    ->group(function () {
        Schedule::command('command-one');
        Schedule::command('command-two');
        Schedule::command('command-three')->everyTenMinutes();
    });
```

Additionally, this implementation allows users to declare schedules similarly to Route, like this:
```php
Schedule::hourly()->command('inspire');
```

File: src/Illuminate/Console/Scheduling/Schedule.php [hunk 4]
```diff
@@ -77,6 +83,13 @@ class Schedule
      */
     protected $mutexCache = [];
 
+    protected ScheduleAttributes $attributes;
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/Schedule.php:86]: Missing doc block.

File: src/Illuminate/Console/Scheduling/Schedule.php [hunk 5]
```diff
@@ -77,6 +83,13 @@ class Schedule
...
+
+    /**
+     * @var array<int, ScheduleAttributes>
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/Schedule.php:89]: Missing a doc block description here (correct the suggested message if it's not suitable):

```suggestion
     * The nested commands to schedule.
     *
     * @var array<int, ScheduleAttributes>
```

File: src/Illuminate/Console/Scheduling/Schedule.php [hunk 6]
```diff
@@ -17,9 +18,14 @@
 use Illuminate\Support\Traits\Macroable;
 use RuntimeException;
 
+/**
+ * @mixin ScheduleAttributes
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/Schedule.php:22]: This should be the full namespace:

```suggestion
 * @mixin \Illuminate\Console\Scheduling\ScheduleAttributes
```

File: src/Illuminate/Console/Scheduling/ManagesAttributes.php [hunk 7]
```diff
@@ -0,0 +1,204 @@
+<?php
+
+declare(strict_types=1);
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/ManagesAttributes.php:3]: Strict types declaration should be removed (not used in the framework), and could also be a breaking change.

Comment (istiak-tridip) [reply src/Illuminate/Console/Scheduling/ManagesAttributes.php:3]: Thanks, @stevebauman! I’ve updated the code.

File: src/Illuminate/Console/Scheduling/Schedule.php [hunk 8]
```diff
@@ -255,9 +275,39 @@ public function exec($command, array $parameters = [])
...
+    }
+
+    /**
+     * Merge the attributes with the given event.
+     */
+    protected function mergeAttributes(Event $event): void
+    {
+        if (isset($this->attributes)) {
+            $this->attributes->mergeAttributes($event);
+            unset($this->attributes);
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/Schedule.php:302]: Nitpick -- spacing

```suggestion
            $this->attributes->mergeAttributes($event);
            
            unset($this->attributes);
```

File: src/Illuminate/Console/Scheduling/Schedule.php [hunk 9]
```diff
@@ -255,9 +275,39 @@ public function exec($command, array $parameters = [])
...
+        }
+
+        if (! empty($this->groupStack)) {
+            $group = end($this->groupStack);
+            $group->mergeAttributes($event);
```
Comment (stevebauman) [inline src/Illuminate/Console/Scheduling/Schedule.php:307]: Nitpick -- unnecessary variable creation:

```suggestion
            end($this->groupStack)->mergeAttributes($event);
```

**Review by stevebauman — COMMENTED** [review]
Just nits left. Everything LGTM! 👍 

We will see what the Laravel team thinks.

File: src/Illuminate/Console/Scheduling/Schedule.php (hunk 9 above)
Comment (istiak-tridip) [reply src/Illuminate/Console/Scheduling/Schedule.php:308]: Thanks! But I prefer the current approach for clarity and debugging, and it’s consistent with similar usage elsewhere in the framework.

Comment (stevebauman) [reply src/Illuminate/Console/Scheduling/Schedule.php:308]: Ok sounds good!

**Comment by taylorotwell** [comment]
Thanks @istiak-tridip!

**Comment by decadence** [comment]
Strange but I get 
```
Undefined property: Illuminate\Console\Scheduling\PendingEventAttributes::$mutex, vendor/laravel/framework/src/Illuminate/Console/Scheduling/ManagesAttributes.php, 145
```

when I use group with `withoutOverlapping` and run `schedule:run`.

```php
Schedule::runInBackground()
    ->withoutOverlapping()
    ->everyMinute()
    ->group(function () {
        Schedule::command(FetchSchedule::class);
    });
```

If I comment `withoutOverlapping` Schedule works.

**Comment by istiak-tridip** [comment]
@decadence I was able to reproduce the issue and have submitted a bug fix in PR #53553. Could you please test the fix and share your feedback on the PR?

Thanks!

**Comment by decadence** [comment]
@istiak-tridip I think I found another bug with Schedule::group.

I have this events:

```php
Schedule::runInBackground()
    ->group(function () {
        Schedule::dailyAt("09:00")
            ->group(function () {
                Schedule::command(ConnectVps::class);
            });

        Schedule::command(ActiveDocuments::class)
            ->everyTenMinutes();
    });
```

As you can see `ActiveDocuments::class` runs not everyTenMinutes but everyTenMinutes only on 9 hour.
image

**Pull request merged**